protocol/forge-scripts-deploy-in-go.md

```diff
@@ -0,0 +1,222 @@
+and write any tests to exercise that deployment, as should be the default for every protocol feature.
+
+The deployment implementation details stay encapsulated in the Forge scripting,
+which is unified with the Forge testing, and thus unifying the production and test code paths.
+
+## Potential future extension: deployment integration
+
+The `vm.broadcast` cheat-code is an opportunity for future deployment improvements:
+rather than running through Forge script when preparing the production deployment transactions,
+we could run through the Go genesis tool.
+
+This would allow us to script more advanced post-processing of the transactions:
+- cross-validation against the superchain-registry
+- simulation of the transaction with custom tracing
+
+And all bundled in Go, so the end-user does not need to ensure a specific Forge version,
+does not need to as many manual `forge script` invocations,
+and all environment settings (that might otherwise be set with forge script flags)
+can be controlled by defining the exact CLI interface.
+
+This is out-of-scope for now, but may help unify the deploy process that production chains use,
+and the deploy process that devnets / op-e2e use.
+
+# Alternatives Considered
+
+See proposed solution [in design doc 52](https://github.com/ethereum-optimism/design-docs/pull/52),
+where `forge script` is used as is, and performance concerns are mitigated with caching.
+
+An [experimental draft of the caching](https://github.com/ethereum-optimism/optimism/pull/11297) was implemented,
+but arguably the caching introduced too much complexity and fragility.
+
+Other solutions / ideas are discussed in design-doc 52 as well, but were not viable.
+
+# Risks & Uncertainties
+
+## Geth Go EVM
+
+The Go EVM instrumentation might be difficult, as the geth EVM is not as widely used in tooling.
+However, the tracing functionality is excellent, and Go is quite flexible.
+If we need to we can make very minor tweaks to `op-geth`,
+to expose any inaccessible EVM internals needed to implement the forge cheatcodes.
+
+## Eng time
+
+This is a mini-project: the scope of implementing the cheat-codes is not that large (a few days at most),
+but the integration into tooling and op-e2e may be more involved (can be a week, maybe two).
+
+In the past the `L2Genesis.s.sol` and `allocs` work, that moved us away from the manual and error-prone op-chain-ops surgery,
+was completed successfully in the form of an interop side-project to remove tech-debt. This project scope looks quite similar.
+
+This functionality does block Interop op-e2e testing: without it,
+we are not able to customize deployments sufficiently and cleanly (avoiding many more `allocs` special cases),
+to get multi-L2 deployments into the op-e2e.
+
+## Devrel feedback
+
+Historically devrel has not been included sufficiently in the deployment-flow design.
+Known pain-points like inconsistency between the Go and forge genesis generation,
```

Comment:
Good call out here, for example, a recent partner was having trouble using the solidity genesis allocs creation on the `op-contracts/v1.3.0` release so they opted for something ~200 commits into the future that played nicely with the tooling instead of using the legacy genesis creation method.

With that being called out, if this lands after a L1 contract release, it would be nice to have this type of functionality cut in a different minor release so its accessible for new chain deployments between say `op-contracts/v1.6.0` and `op-contracts/v1.7.0`.